encoders/firebase-decoders-json/src/json/java/com/google/firebase/decoders/json/JsonDataDecoderBuilderContext.java

```diff
@@ -0,0 +1,119 @@
+  private JsonReader reader;
+
+  public JsonDataDecoderBuilderContext(@NonNull Map<TypeToken.ClassToken<?>, ObjectDecoder<?>> objectDecoders) {
+    for (Map.Entry<TypeToken.ClassToken<?>, ObjectDecoder<?>> entry : objectDecoders.entrySet()) {
+      TypeToken.ClassToken<?> classToken = entry.getKey();
+      ObjectDecoder objectDecoder = entry.getValue();
+      ObjectDecoderContextImpl objectDecoderCtx = ObjectDecoderContextImpl.of(classToken);
+      objectDecoder.decode(objectDecoderCtx);
+      TypeCreator<?> creator = objectDecoder.decode(objectDecoderCtx);
+      objectDecodersCtx.put(classToken, objectDecoderCtx);
+      typeCreators.put(classToken, creator);
+    }
+  }
+
+  @NonNull
+  @Override
+  public <T> T decode(@NonNull InputStream input, @NonNull TypeToken<T> typeToken) throws IOException {
+    this.reader = new JsonReader(new InputStreamReader(input, "UTF-8"));
+    return decode(typeToken);
+  }
+
+  private <T> T decode(TypeToken<T> typeToken) throws IOException {
+    if (typeToken instanceof TypeToken.ClassToken) {
+      TypeToken.ClassToken classToken = (TypeToken.ClassToken<T>) typeToken;
+      T val = (T) decodeObject(classToken);
+      return val;
+    } else if (typeToken instanceof TypeToken.ArrayToken) {
+      //TODO: Change typeParameter T in ArrayToken<T> to represent component type.
+      /**
+       * reader.beginArray();
+       * List<T> l = new LinkedList<>();
+       * while reader.hasNext:
+       *  T val = decode(arrayToken.getComponentType());
+       *  l.add(val);
+       * reader.endArray();
+       * return l.toArray();
+       */
+    }
+    return null;
+  }
+
+  private <T> T decodeObject(TypeToken.ClassToken<T> classToken) throws IOException {
+    if(classToken.getRawType().isPrimitive() || classToken.getRawType().equals(String.class))
+      return decodePrimitives(classToken.getRawType());
+
+    CreationContextImpl<T> creationCtx = new CreationContextImpl<>();
+    ObjectDecoderContextImpl<T> decoderCtx = (ObjectDecoderContextImpl<T>) objectDecodersCtx.get(classToken);
+    reader.beginObject();
+    while (reader.hasNext()) {
+      String fieldName = reader.nextName();
+      FieldRef<?> fieldRef = decoderCtx.getFieldRef(fieldName);
+      // TODO: fieldRef is null (ObjectDecoder didn't handle that field)
+      if (fieldRef instanceof FieldRef.Primitive) {
+        creationCtx.put(fieldRef,
+                decodePrimitives(((FieldRef.Primitive<?>) fieldRef).getRawType()));
+      } else if (fieldRef instanceof FieldRef.Boxed) {
+        creationCtx.put(fieldRef,
+                decode(((FieldRef.Boxed<?>) fieldRef).getTypeToken()));
+      }
+    }
+    reader.endObject();
+
+    TypeCreator<T> creator = (TypeCreator<T>) typeCreators.get(classToken);
+    return (T) creator.create(creationCtx);
+  }
+
+  private <T> T decodePrimitives(Class<T> clazz) throws IOException {
+    //TODO: try catch
+    if (clazz.equals(Boolean.class)) {
```

Comment:
nit: `Boolean.class` is not primitive, `boolean.class` is. same for `Integer` vs `int`, etc.